### ISSUE TITLE
Graceful Shutdown of Server

### ISSUE THREAD

**Issue by theduke** [bluejekyll/trust-dns#1869]
Am I missing a way to do orderly shutdown of the server?

As in: a way to stop the server from accepting new connections, finish the running requests and then resolve the `ServerFuture`.?

**Comment by djc** [comment]
Yup, it looks like there's currently no way to do that. Would you be able to submit a PR to fix that?

**Comment by bluejekyll** [comment]
Yeah, this hasn't been implemented yet. I've had an old issue from the beginning of the project to implement something else for this as well for doing a config reload on a SIGHUP, #20.

I think this issue could be good for library support, while the SIGHUP would be specific to the `trust-dns` binary. The way I see it is we need a couple of things to be implemented. Internally the server needs to have a shared atomic for triggering the safe shutdown. This would then be wired into all the loops on all the listening sockets to shut down those listeners. The challenge that we'd need a good solution for is to make sure those listening loops are woken even if no new connections come in. I don't know how else to trigger that. There may be some features in Tokio or Futures that I'm unfamiliar with that make that easy.

**Comment by djc** [comment]
Could probably do a `tokio::select!()` against the listener on one side and a sync primitive like a `tokio::sync::Notify` on the other side?

**Comment by bluejekyll** [comment]
Yes; Notify is probably a good option to do that.

**Comment by yerke** [comment]
I think this issue was solved in https://github.com/bluejekyll/trust-dns/pull/1977
